Source/Core/Core/AchievementManager.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2023 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#ifdef USE_RETRO_ACHIEVEMENTS
+
+#include "AchievementManager.h"
+#include "rcheevos/include/rc_api_runtime.h"
+#include "rcheevos/include/rc_api_user.h"
+#include "rcheevos/include/rc_hash.h"
+#include "rcheevos/include/rc_runtime.h"
+
+#include "Common/HttpRequest.h"
+#include "Config/AchievementSettings.h"
+#include "Core/Core.h"
+
+namespace Achievements
+{
+static rc_runtime_t runtime{};
+static bool is_runtime_initialized = false;
+static rc_api_login_response_t login_data{.response{.succeeded = 0}};
+
+namespace  // Hide from use outside this file
+{
+template <typename RcRequest, typename RcResponse>
+void Request(RcRequest rc_request, RcResponse* rc_response,
+             int (*init_request)(rc_api_request_t* request, const RcRequest* api_params),
+             int (*process_response)(RcResponse* response, const char* server_response))
+{
+  rc_api_request_t api_request;
+  Common::HttpRequest http_request;
+  init_request(&api_request, &rc_request);
+  auto http_response = http_request.Post(api_request.url, api_request.post_data);
+  if (http_response.has_value() && http_response.value().size() > 0)
+  {
+    char* response_str = (char*)calloc(http_response.value().size() + 1, 1);
```

Comment:
Is there any reason to not just use a std::string here? Doesn't seem like process_response does anything with the allocated pointer itself, considering it's freed below.

Reply:
Actually, looking at this again, shouldn't it be sufficient to get rid of the loop and allocation entirely and collapse the if statement body down to just:

```cpp
if (http_response && !http_response->empty())
{
  process_response(rc_response, http_response->c_str());
}
```

since `c_str()` guarantees null termination?

Reply:
process_response is a function pointer passed into this method. It might be one of a number of rc_api_process_<x>_response methods, for example https://github.com/RetroAchievements/rcheevos/wiki/rc_api_process_login_response , which accept the second parameter as a c-string input and use that input to populate the response struct passed in as a pointer to the first parameter. So yes, response_str is used to generate rc_response and then it is no longer needed.

Reply:
Right, but you can get a C string out of a std::string in the manner I described without needing to do unnecessary manual memory management. I was just asking why that wasn't used instead.

Reply:
No worries, it happens